File: apps/api/src/services/cosine-select.ts
import embeddingsAsset from "./exercise_embeddings.json";
import manifest from "./exercise_primitives_manifest.json";

// Cosine drill selection: rank catalog drills by similarity of their Aria
// embedding to a query embedding of the student's weak passage, WITHIN the
// teacher-diagnosed dimension bucket (FILTER then RANK). This replaces the
// constant-per-dimension top-1 selector when a query embedding is available;
// when it is not (no AMT MIDI for the weak bars), callers fall back to the
// deterministic selectPrimitive path.
//
// The catalog vectors in exercise_embeddings.json are L2-normalized at export,
// so cosine reduces to a dot product once the query is normalized here.

type EmbeddingsAsset = { dim: number; ids: string[]; vectors: number[][] };
type ManifestEntry = { dimensions: string[] };

const ASSET = embeddingsAsset as EmbeddingsAsset;
const MANIFEST = manifest as Record<string, ManifestEntry>;

// id -> normalized vector, built once at module load.
const VECTORS: Map<string, number[]> = new Map(
  ASSET.ids.map((id, i) => [id, ASSET.vectors[i]]),
);

export type CosineMatch = { primitiveId: string; score: number };

function l2normalize(v: number[]): number[] {
  let sumsq = 0;
  for (const x of v) sumsq += x * x;
  const norm = Math.sqrt(sumsq);
  if (norm === 0) {
    // A zero query has no direction; refuse it rather than dividing by zero.
    throw new Error("cosine-select: query embedding has zero magnitude");
  }
  return v.map((x) => x / norm);
}

function dot(a: number[], b: number[]): number {
  let s = 0;
  for (let i = 0; i < a.length; i++) s += a[i] * b[i];
  return s;
}

/**
 * Rank catalog drills tagged for `targetDimension` by cosine similarity to
 * `query`, returning the best match. Returns null when no catalog drill is
 * tagged for the dimension (the caller widens to the neutral default), mirroring
 * the empty-bucket branch of selectPrimitive.
 *
 * Ties (equal cosine) break by primitive_id ascending, matching the model-side
 * match_exercises determinism.
 */
export function cosineSelectWithinDimension(
  query: number[],
  targetDimension: string,
): CosineMatch | null {
  if (query.length !== ASSET.dim) {
    throw new Error(
      `cosine-select: query dim ${query.length} != catalog dim ${ASSET.dim}`,
    );
  }
  const q = l2normalize(query);

  let best: CosineMatch | null = null;
  for (const id of ASSET.ids) {
    const entry = MANIFEST[id];
    if (entry === undefined || !entry.dimensions.includes(targetDimension)) {
      continue;
    }
    const vec = VECTORS.get(id);
    if (vec === undefined) continue;
    const score = dot(q, vec);
    if (
      best === null ||
      score > best.score ||
      (score === best.score && id < best.primitiveId)
    ) {
      best = { primitiveId: id, score };
    }
  }
  return best;
}

File: apps/evals/pipeline/exercise_routing/cosine_select.py
"""Python mirror of cosine-select.ts for the eval's cosine A/B re-measure.

The eval compares selector strategies on the SAME sessions: the deterministic
top-1 (selection.select_primitive) vs cosine-within-dimension (here). Both read
the same committed catalog asset (exercise_embeddings.json, L2-normalized) so the
eval's cosine choice matches what the Worker's cosine-select.ts would pick for the
same query -- a faithful counterfactual, not a re-implementation that could drift.
"""
from __future__ import annotations

import json
from pathlib import Path

import numpy as np

_DEFAULT_ASSET = (
    Path(__file__).resolve().parents[3]
    / "api" / "src" / "services" / "exercise_embeddings.json"
)


class CatalogEmbeddings:
    """Preloaded L2-normalized catalog matrix + id list + manifest dimensions."""

    def __init__(self, asset_path: Path, manifest: dict):
        data = json.loads(Path(asset_path).read_text())
        self.dim: int = data["dim"]
        self.ids: list[str] = data["ids"]
        self.matrix = np.asarray(data["vectors"], dtype=np.float32)  # (n, dim), normalized
        if self.matrix.shape != (len(self.ids), self.dim):
            raise ValueError(
                f"asset matrix {self.matrix.shape} != ({len(self.ids)}, {self.dim})"
            )
        self.manifest = manifest


def load_catalog(manifest: dict, asset_path: Path = _DEFAULT_ASSET) -> CatalogEmbeddings:
    return CatalogEmbeddings(asset_path, manifest)


def cosine_select_within_dimension(
    query: np.ndarray, target_dimension: str, catalog: CatalogEmbeddings
) -> str | None:
    """Return the primitive_id best matching `query` among drills tagged for the
    dimension, or None if the bucket is empty (caller widens). Ties break by id.
    """
    q = np.asarray(query, dtype=np.float32)
    if q.shape != (catalog.dim,):
        raise ValueError(f"query dim {q.shape} != catalog dim ({catalog.dim},)")
    norm = float(np.linalg.norm(q))
    if norm == 0:
        raise ValueError("query embedding has zero magnitude")
    q = q / norm

    sims = catalog.matrix @ q  # (n,) cosine (matrix pre-normalized)
    best_id: str | None = None
    best_score = -2.0
    for i, pid in enumerate(catalog.ids):
        entry = catalog.manifest.get(pid)
        if entry is None or target_dimension not in entry.get("dimensions", []):
            continue
        s = float(sims[i])
        if s > best_score or (s == best_score and (best_id is None or pid < best_id)):
            best_score = s
            best_id = pid
    return best_id
